### ISSUE TITLE
Meridian Convergence and Scale Factor

### ISSUE THREAD

**Issue by wg1562** [pyproj4/pyproj#503]
Hi All,

I joined GitHub to get help with this issue, related to:
https://github.com/pyproj4/pyproj/issues/144

I want to use pyproj in python to convert coordinates and get the meridian convergence angle and grid scale factor. I figured out from the proj command line to use the -V flag to get these values, but it would be great to be able to get these values directly with pyproj in python.

There was talk of adding this feature into pyproj, but snooping through the scripts I didn't see anything. Is there any plans of still adding it? 

Thanks.

**Comment by snowman2** [comment]
Thanks for bringing this up. Looks like it got lost. Sounds useful to add.

Do you have an example of how you use this functionality?

**Comment by snowman2** [comment]
PROJ API:

https://github.com/OSGeo/PROJ/blob/5be02bf6d1d45ac0d835c138074611843982e9e8/src/proj.h#L170-L184

https://github.com/OSGeo/PROJ/blob/5be02bf6d1d45ac0d835c138074611843982e9e8/src/proj.h#L437

**Comment by wg1562** [comment]
I think it would just be great to elevate the proj_factors() function into pyproj.py. I don't know how hard this would be to implement, but I think it would be easier than editing the proj() function itself.

Thanks.

**Comment by snowman2** [comment]
You can try it out by:

```
pip install git+git://github.com/pyproj4/pyproj.git@0910f97#egg=pyproj
```

If you need to install PROJ 6.2+ beforehand, I would recommend conda:
```
conda create -n proj -c conda-forge proj cython
conda activate proj
pip install git+git://github.com/pyproj4/pyproj.git@0910f97#egg=pyproj
```

**Comment by wg1562** [comment]
I was able to install PROJ 6.3.1 with no problems, however, when I ran:

pip install git+git://github.com/pyproj4/pyproj.git@0910f97#egg=pyproj

I got:

"WARNING: Did not find branch or tag '0910f97', assuming revision or ref."

Then it just crashed. I apologize if I'm just missing something basic. I appreciate all of your help with this.

-wg

**Comment by snowman2** [comment]
> "WARNING: Did not find branch or tag '0910f97', assuming revision or ref."

That's okay - it is working as expected.

How did you install PROJ 6.3.1? What version of python? What version of pip?

This is what it looked like for me:
```
(base)$ conda create -n proj -c conda-forge proj cython
Collecting package metadata (current_repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.7.12
  latest version: 4.8.2

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/snowal/miniconda/envs/proj

  added / updated specs:
    - cython
    - proj


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-0_gnu
  ca-certificates    conda-forge/linux-64::ca-certificates-2019.11.28-hecc5488_0
  certifi            conda-forge/linux-64::certifi-2019.11.28-py38_0
  cython             conda-forge/linux-64::cython-0.29.15-py38he1b5a44_0
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.33.1-h53a641e_8
  libffi             conda-forge/linux-64::libffi-3.2.1-he1b5a44_1006
  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.2.0-h24d8f2e_2
  libgomp            conda-forge/linux-64::libgomp-9.2.0-h24d8f2e_2
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.2.0-hdf63c60_2
  ncurses            conda-forge/linux-64::ncurses-6.1-hf484d3e_1002
  openssl            conda-forge/linux-64::openssl-1.1.1d-h516909a_0
  pip                conda-forge/noarch::pip-20.0.2-py_2
  proj               conda-forge/linux-64::proj-6.3.1-hc80f0dc_0
  python             conda-forge/linux-64::python-3.8.1-h357f687_2
  readline           conda-forge/linux-64::readline-8.0-hf8c457e_0
  setuptools         conda-forge/linux-64::setuptools-45.2.0-py38_0
  sqlite             conda-forge/linux-64::sqlite-3.30.1-hcee41ef_0
  tk                 conda-forge/linux-64::tk-8.6.10-hed695b0_0
  wheel              conda-forge/noarch::wheel-0.34.2-py_1
  xz                 conda-forge/linux-64::xz-5.2.4-h14c3975_1001
  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1006


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate proj
#
# To deactivate an active environment, use
#
#     $ conda deactivate

(base) $ conda activate proj
(proj) $ pip install git+git://github.com/pyproj4/pyproj.git@0910f97#egg=pyproj
Collecting pyproj
  Cloning git://github.com/pyproj4/pyproj.git (to revision 0910f97) to /tmp/pip-install-z8zvz9t4/pyproj
  Running command git clone -q git://github.com/pyproj4/pyproj.git /tmp/pip-install-z8zvz9t4/pyproj
  WARNING: Did not find branch or tag '0910f97', assuming revision or ref.
  Running command git checkout -q 0910f97
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: pyproj
  Building wheel for pyproj (PEP 517) ... done
  Created wheel for pyproj: filename=pyproj-2.5.0-cp38-cp38-linux_x86_64.whl size=1536211 sha256=630b55ae3174d9a6a18ed260127ba2533b91dc48ca1a2ce937c11edc39d8e296
  Stored in directory: /tmp/pip-ephem-wheel-cache-u8px_o8y/wheels/8d/da/2d/87504e818817bfc82f1e29e05b16c551ee5fa913e253c2b0f4
Successfully built pyproj
Installing collected packages: pyproj
Successfully installed pyproj-2.5.0
```

**Comment by wg1562** [comment]
I am using python 3.7.4 and pip 19.3.1. I installed PROJ 6.3.1 using "conda create -n proj -c conda-forge proj cython"

Then when I try "pip install git+git://github.com/pyproj4/pyproj.git@0910f97#egg=pyproj
Collecting pyproj"
I get:

```
 Building wheel for pyproj (PEP 517) ... error
ERROR: Command errored out with exit status 1:
... (lots of stuff here)
  error: Microsoft Visual C++ 14.0 is required. Get it with "Build Tools for Visual Studio": https://visualstudio.microsoft.com/downloads/
  ----------------------------------------
  ERROR: Failed building wheel for pyproj
Failed to build pyproj
ERROR: Could not build wheels for pyproj which use PEP 517 and cannot be installed directly
```

I looked into installing : Microsoft Visual C++ 14.0, but it's 6.5 GB and I don't have that much space left on my drive. 

Thanks

**Comment by snowman2** [comment]
I wonder if you could install python 3.6 and try this wheel? https://ci.appveyor.com/api/buildjobs/lyqh293gwvuxg5is/artifacts/dist%2Fpyproj-2.5.0-cp36-cp36m-win_amd64.whl

It might not work, but it probably won't hurt to try.

**Comment by wg1562** [comment]
I tried uninstalling and re-installing an old 3.6 miniconda. I pip installed your wheel and then pyproj isn't recognized (but bit is installed strangely. I can see the folder). So I ran the install on proj 6.3 and then pip install git+git://github.com/pyproj4/pyproj.git@0910f97#egg=pyproj

Same errors as before. I appreciate all your help with this, but I have to admit python is not my forte and I'm probably missing something obvious. You don't need to keep spending time on this. 

Thanks.

**Comment by snowman2** [comment]
Okay, I guess you will need to wait until the next release.

**Comment by wg1562** [comment]
snowman,

I installed 2.6.0 think I got .get_factors() to work. However, if I submit coordinates (lam,phi) as decimal degrees, I get an "latitude or longitude exceeded limits" error. If I convert coordinates to radians, I get results, but they are far off from what is expected. Am I using the function correctly? Please see code below.

Thanks.

```
lam = -120 
phi = 34   
lam_r = lam * math.pi/180
phi_r = phi * math.pi/180   
#epsg='2770' #2770 is CA 5 ZONE SPC
epsg = '3717'  #3717 is UTM 10N for WGS84
crs_data = CRS.from_epsg(epsg)
print (crs_data.get_geod)
epsg_str = 'EPSG:'+ epsg
proj_used = pyproj.Proj(epsg_str) 
x,y = proj_used(lam, phi)
print(lam, phi)
print(lam_r, phi_r)
print (x, y)
factors = proj_used.get_factors(lam,phi,False,True)
print (factors)
#Results from NCAT:
#https://www.ngs.noaa.gov/NCAT/
#LLH = (-120,34,100)  #test coordinates
#epsg = 3717 is UTM 10N for WGS84
#Northing (m)= 3,766,214.716
#Easting (m)= 777,091.295
#convDMS = 01 40 43.13
#convRAD = 0.02929792
#gridSF = 1.00054660
```

**Comment by snowman2** [comment]
@wg1562 mind opening up a new issue for this?

**Comment by kimollivier** [comment]
Successfully built pyproj
Installing collected packages: pyproj
Successfully installed pyproj-2.5.0
...after installing VC++ Build Tools
I want to do the same thing: get the convergence at the centre of each map to make up a nice compass in the margin with convergence and magnetic deviation. Looking promising.

**Comment by snowman2** [comment]
You should have better success with `pyproj==2.6.1.post1`

**Comment by kimollivier** [comment]
Ok, I updated to pyproj 2.6.1.post1. That fixed the bug requiring radians only. ;-)
I tested out the tool (why not?) using the Esri Tool for an independent calculation and the results agree to 5 dec places.
The only worry is the signs are opposite. Maybe this is just the point of view between the definition of True North from Grid North and Meridian_Convergence?
I used NZTM (epsg:2193) for my test, which is in the Southern Hemisphere, always an untested area!
![sample](https://user-images.githubusercontent.com/3343876/81869191-b96a3f80-95c7-11ea-87a1-b1b980e77ef3.png)
So a great new function that I could not find anywhere else in FOSS4G tools. Thanks so much.
